File: solutions/javascript/affine-cipher/1/affine-cipher.js
const ALPHABET = 'abcdefghijklmnopqrstuvwxyz';
const ALPHABET_LENGTH = ALPHABET.length;

/**
 * Checks if two numbers are coprimes.
 * @param {number} a - First number.
 * @param {number} b - Second number.
 * @returns {boolean} True if a and b are coprimes, otherwise false.
 */
const areCoprimes = (a, b) => {
  for (let i = Math.min(a, b); i > 1; i--) {
    if (a % i === 0 && b % i === 0) {
      return false;
    }
  }
  return true;
};

/**
 * Ensures that two numbers are coprimes.
 * @param {number} a - First number.
 * @param {number} b - Second number.
 * @throws {Error} If a and b are not coprimes.
 */
const checkCoprime = (a, b) => {
  if (!areCoprimes(a, b)) {
    throw new Error('a and m must be coprime.');
  }
};

/**
 * Checks if a value is a number.
 * @param {any} candidate - The value to check.
 * @returns {boolean} True if the candidate is a number, otherwise false.
 */
const isNumber = (candidate) => !isNaN(Number(candidate));

/**
 * Finds the Modular Multiplicative Inverse (MMI) of a number.
 * @param {number} a - The number to find the MMI for.
 * @returns {number} The MMI of a.
 */
const findMMI = (a) => {
  let i = 1;
  while (true) {
    i++;
    if ((a * i - 1) % ALPHABET_LENGTH === 0) {
      return i;
    }
  }
};

/**
 * Computes the positive modulo of two numbers.
 * @param {number} a - The dividend.
 * @param {number} b - The divisor.
 * @returns {number} The positive modulo of a and b.
 */
const positiveModulo = (a, b) => ((a % b) + b) % b;

/**
 * Groups elements into sub-arrays of specified length.
 * @param {Array} elements - The elements to group.
 * @param {number} groupLength - The length of each group.
 * @returns {Array} The grouped elements.
 */
const groupBy = (elements, groupLength) => {
  const result = [[]];
  let i = 0;
  elements.forEach((el) => {
    if (result[i] && result[i].length < groupLength) {
      result[i].push(el);
    } else {
      i++;
      result[i] = [el];
    }
  });
  return result;
};

/**
 * Encodes a phrase using the affine cipher.
 * @param {string} phrase - The phrase to encode.
 * @param {Object} keys - The keys for the cipher.
 * @param {number} keys.a - The 'a' key.
 * @param {number} keys.b - The 'b' key.
 * @returns {string} The encoded phrase.
 */
export const encode = (phrase, { a, b }) => {
  checkCoprime(a, ALPHABET_LENGTH);
  let encodedText = '';

  phrase
    .toLowerCase()
    .split('')
    .filter((char) => char !== ' ')
    .forEach((char) => {
      if (ALPHABET.includes(char)) {
        const x = ALPHABET.indexOf(char);
        const encodedIndex = (a * x + b) % ALPHABET_LENGTH;
        encodedText += ALPHABET[encodedIndex];
      } else if (isNumber(char)) {
        encodedText += char;
      }
    });

  return groupBy(encodedText.split(''), 5)
    .map((group) => group.join(''))
    .join(' ');
};

/**
 * Decodes a phrase using the affine cipher.
 * @param {string} phrase - The phrase to decode.
 * @param {Object} keys - The keys for the cipher.
 * @param {number} keys.a - The 'a' key.
 * @param {number} keys.b - The 'b' key.
 * @returns {string} The decoded phrase.
 */
export const decode = (phrase, { a, b }) => {
  checkCoprime(a, ALPHABET_LENGTH);
  const mmi = findMMI(a);

  return phrase
    .split('')
    .filter((char) => char !== ' ')
    .map((char) => {
      if (isNumber(char)) {
        return char;
      }
      const y = ALPHABET.indexOf(char);
      const decodedIndex = positiveModulo(mmi * (y - b), ALPHABET_LENGTH);
      return ALPHABET[decodedIndex];
    })
    .join('');
};

File: solutions/python/affine-cipher/3/affine_cipher.py
from string import ascii_lowercase
from typing import List

ALPHABET = ascii_lowercase
TO_REMOVE = (" ", ".", ",")


def hcf_check(x: int, y: int) -> int:
    """Calculate the highest common factor (HCF) of two integers x and y using recursion.

    Args:
        x (int): First integer.
        y (int): Second integer.

    Returns:
        int: The highest common factor of x and y.
    """
    if y == 0:
        return x
    return hcf_check(y, x % y)


def mmi(a: int) -> int:
    """Find the modular multiplicative inverse of integer a under modulo 26.

    Args:
        a (int): The integer to find the modular inverse for.

    Returns:
        int: The modular multiplicative inverse of a.
    """
    for x in range(1, 26):
        if (a * x) % 26 == 1:
            return x


def encode(plain_text: str, a: int, b: int) -> str:
    """Encode the given plain_text using the affine cipher formula defined by keys a and b.

    Args:
        plain_text (str): The text to encode.
        a (int): The first key for encoding.
        b (int): The second key for encoding.

    Returns:
        str: The encoded text.

    Raises:
        ValueError: If a and m are not coprime.
    """
    if hcf_check(26, a) != 1:
        raise ValueError("a and m must be coprime.")
    for character in TO_REMOVE:
        plain_text = plain_text.replace(character, "")

    combined = ""

    for i in plain_text.lower():
        if i in ALPHABET:
            ind = (a * ALPHABET.index(i) + b) % 26
            combined += ALPHABET[ind]
        else:
            combined += i

    combined_split: List[str] = [
        combined[k : k + 5] for k in range(0, len(combined), 5)
    ]
    return " ".join(combined_split)


def decode(ciphered_text: str, a: int, b: int) -> str:
    """Decode the given ciphered_text using the affine cipher formula defined by keys a and b.

    Args:
        ciphered_text (str): The text to decode.
        a (int): The first key for decoding.
        b (int): The second key for decoding.

    Returns:
        str: The decoded text.

    Raises:
        ValueError: If a and m are not coprime.
    """
    if hcf_check(26, a) != 1:
        raise ValueError("a and m must be coprime.")
    ciphered_text = ciphered_text.replace(" ", "")
    result = ""
    for character in ciphered_text:
        if character in ALPHABET:
            ind = ((mmi(a)) * (ALPHABET.index(character) - b)) % 26
            result += ALPHABET[ind]
        else:
            result += character
    return result
